src/Lean/Eval.lean

```diff
@@ -3,24 +3,62 @@ Copyright (c) 2019 Microsoft Corporation. All rights reserved.
 Released under Apache 2.0 license as described in the file LICENSE.
 Authors: Leonardo de Moura, Sebastian Ullrich
 -/
-import Lean.Environment
+import Lean.Elab.Command
 
 namespace Lean
 
-universe u
+open Elab.Command Elab.Term
 
 /--
 `Eval` extension that gives access to the current environment & options.
 The basic `Eval` class is in the prelude and should not depend on these
 types.
 -/
 class MetaEval (α : Type u) where
-  eval : Environment → Options → α → (hideUnit : Bool) → IO Environment
+  eval : α → (hideUnit : Bool) → CommandElabM (Option MessageData)
 
-instance {α : Type u} [Eval α] : MetaEval α :=
-  ⟨fun env _ a hideUnit => do Eval.eval (fun _ => a) hideUnit; pure env⟩
+instance [Eval α] : MetaEval α where
+  eval a hideUnit := return ← Eval.eval (hideUnit := hideUnit) fun _ => a
 
-def runMetaEval {α : Type u} [MetaEval α] (env : Environment) (opts : Options) (a : α) : IO (String × Except IO.Error Environment) :=
-  IO.FS.withIsolatedStreams (MetaEval.eval env opts a false |>.toBaseIO)
+instance [ToMessageData α] : MetaEval α where
+  eval a _ := return toMessageData a
+
+instance : MetaEval Unit where
+  eval _ hideUnit := return if hideUnit then none else some "()"
+
+instance [MetaEval α] : MetaEval (BaseIO α) where
+  eval a _ := do MetaEval.eval (hideUnit := true) (← liftCoreM a)
+
+instance [MetaEval α] : MetaEval (IO α) where
+  eval a _ := do MetaEval.eval (hideUnit := true) (← liftCoreM a)
```

Comment:
Not that it really matters but I think this could be lifted directly
```suggestion
  eval a _ := do MetaEval.eval (hideUnit := true) (← liftIO a)

instance [MetaEval α] : MetaEval (IO α) where
  eval a _ := do MetaEval.eval (hideUnit := true) (← liftIO a)
```